File: test/ModularPipelines.UnitTests/Helpers/SerializationTestModels.cs
using System.Text.Json.Serialization;

namespace ModularPipelines.UnitTests.Helpers;

/// <summary>
/// Shared test models for serialization tests (JSON, XML, YAML).
/// Reduces duplication of similar model definitions across test files.
/// </summary>
public static class SerializationTestModels
{
    /// <summary>
    /// Standard test values used across serialization tests.
    /// </summary>
    public static class TestValues
    {
        public const string FooValue = "Bar!";
        public const string HelloValue = "World!";
        public static readonly string[] ItemsValue = ["One", "Two", "3"];
    }

    /// <summary>
    /// Base test model with common properties for serialization testing.
    /// </summary>
    public record SerializationTestModel
    {
        public string? Foo { get; set; }
        public string? Hello { get; set; }
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public List<string>? Items { get; set; }

        /// <summary>
        /// Creates a new model with default test values.
        /// </summary>
        public static SerializationTestModel CreateDefault() =>
            new() { Foo = TestValues.FooValue, Hello = TestValues.HelloValue };

        /// <summary>
        /// Creates a new model with default test values including items.
        /// </summary>
        public static SerializationTestModel CreateWithItems() =>
            new()
            {
                Foo = TestValues.FooValue,
                Hello = TestValues.HelloValue,
                Items = [..TestValues.ItemsValue],
            };
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Helpers/SerializationTestModels.fs
namespace ModularPipelines.UnitTests.FSharp.Helpers

open System.Collections.Generic
open System.Text.Json.Serialization

/// <summary>
/// Shared test models for serialization tests (JSON, XML, YAML).
/// Reduces duplication of similar model definitions across test files.
/// </summary>
module SerializationTestModels =
    /// <summary>
    /// Standard test values used across serialization tests.
    /// </summary>
    module TestValues =
        let FooValue = "Bar!"
        let HelloValue = "World!"
        let ItemsValue = [| "One"; "Two"; "3" |]

    /// <summary>
    /// Base test model with common properties for serialization testing.
    /// </summary>
    [<CLIMutable>]
    type SerializationTestModel =
        {
            Foo: string
            Hello: string
            [<JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)>]
            Items: List<string>
        }
        
        /// <summary>
        /// Creates a new model with default test values.
        /// </summary>
        static member CreateDefault() =
            { Foo = TestValues.FooValue
              Hello = TestValues.HelloValue
              Items = null }

        /// <summary>
        /// Creates a new model with default test values including items.
        /// </summary>
        static member CreateWithItems() =
            { Foo = TestValues.FooValue
              Hello = TestValues.HelloValue
              Items = List<string>(TestValues.ItemsValue) }
